### PR TITLE
Working video slider & playback speed

### PR DESCRIPTION
**Pull request by fredrikiselius**
* Added slider functionality
* Video playback now returns to start after reaching the end of the clip
* Added functionality for playback speed

### CODE REVIEW

File: ViAn/GUI/mainwindow.cpp [hunk 1]
```diff
@@ -232,25 +237,55 @@ void MainWindow::resizeEvent(QResizeEvent* event) {
...
 void MainWindow::on_videoSlider_valueChanged(int newPos){
+    changeCounter++;
+    bool was_paused = false;
     // Make slider to follow the mouse directly and not by pageStep steps
+    if (abs(prev_slider_pos - newPos) == 1) {
+        return;
+    } else if (slider_moving) {
+        QPoint localMousePos = ui->videoSlider->mapFromGlobal(QCursor::pos());
+        float posRatio = localMousePos.x() / (float )ui->videoSlider->size().width();
+        int sliderRange = ui->videoSlider->maximum() - ui->videoSlider->minimum();
```
Comment (henpe071) [inline ViAn/GUI/mainwindow.cpp]: snake_case

File: ViAn/GUI/mainwindow.cpp [hunk 2]
```diff
@@ -232,25 +237,55 @@ void MainWindow::resizeEvent(QResizeEvent* event) {
...
+        int sliderPosUnderMouse = ui->videoSlider->minimum() + sliderRange * posRatio;
+
+        /*
+        if (!mvideo_player->is_stopped() && !mvideo_player->is_paused()) {
+            std::cout << "pausing video" << std::endl;
+            emit set_pause_video();
+            was_paused = true;
+        }*/
```
Comment (parso619) [inline ViAn/GUI/mainwindow.cpp]: Why is this out-commented code here?

File: ViAn/GUI/mainwindow.cpp [hunk 3]
```diff
@@ -141,7 +143,8 @@ void MainWindow::on_playPauseButton_clicked() {
  * The button supposed to play the video faster
```
Comment (parso619) [inline ViAn/GUI/mainwindow.cpp]: Remove "suppose" in the comment

File: ViAn/GUI/mainwindow.cpp [hunk 4]
```diff
@@ -111,7 +113,7 @@ void MainWindow::set_status_bar(string status, int timer){
  *
```
Comment (parso619) [inline ViAn/GUI/mainwindow.cpp]: Remove "suppose" in the comment

File: ViAn/GUI/mainwindow.cpp [hunk 5]
```diff
@@ -111,7 +113,7 @@ void MainWindow::set_status_bar(string status, int timer){
...
  */
 void MainWindow::on_fastBackwardButton_clicked(){
-
+    mvideo_player->dec_playback_speed();
```
Comment (parso619) [inline ViAn/GUI/mainwindow.cpp]: Add status bar message

File: ViAn/GUI/mainwindow.cpp [hunk 6]
```diff
@@ -232,25 +237,55 @@ void MainWindow::resizeEvent(QResizeEvent* event) {
...
     bool clickOnSlider = (btns & Qt::LeftButton) &&
                          (localMousePos.x() >= 0 && localMousePos.y() >= 0 &&
                           localMousePos.x() < ui->videoSlider->size().width() &&
                           localMousePos.y() < ui->videoSlider->size().height());
-    if (clickOnSlider)
-    {
+    if (clickOnSlider) {
         // Attention! The following works only for Horizontal, Left-to-right sliders
         float posRatio = localMousePos.x() / (float )ui->videoSlider->size().width();
         int sliderRange = ui->videoSlider->maximum() - ui->videoSlider->minimum();
```
Comment (henpe071) [inline ViAn/GUI/mainwindow.cpp]: case

File: ViAn/GUI/mainwindow.cpp [hunk 7]
```diff
@@ -141,7 +143,8 @@ void MainWindow::on_playPauseButton_clicked() {
  * The button supposed to play the video faster
  */
 void MainWindow::on_fastForwardButton_clicked(){
-
+    mvideo_player->inc_playback_speed();
+    set_status_bar("Increased playback speed");
```
Comment (parso619) [inline ViAn/GUI/mainwindow.cpp]: Maybe add the speed to the status bar message.

File: ViAn/GUI/mainwindow.cpp [hunk 8]
```diff
@@ -766,3 +800,34 @@ void MainWindow::enable_video_buttons() {
     ui->previousFrameButton->setEnabled(true);
     ui->stopButton->setEnabled(true);
 }
+
+/**
+ * @brief MainWindow::on_videoSlider_sliderPressed
+ * Block slider update from video_player
+ */
+void MainWindow::on_videoSlider_sliderPressed() {
+    std::cout << "Slider pressed" << std::endl;
```
Comment (parso619) [inline ViAn/GUI/mainwindow.cpp]: Are these prints needed?

File: ViAn/GUI/mainwindow.cpp [hunk 9]
```diff
@@ -766,3 +800,34 @@ void MainWindow::enable_video_buttons() {
...
+        emit set_pause_video();
+    }
+}
+
+/**
+ * @brief MainWindow::on_videoSlider_sliderReleased
+ * Set video playback pos to slider pos and unblock slider update
+ */
+void MainWindow::on_videoSlider_sliderReleased() {
+    std::cout << "Slider released " << mvideo_player->is_paused() << std::endl;
```
Comment (parso619) [inline ViAn/GUI/mainwindow.cpp]: Are these prints needed?

File: ViAn/GUI/mainwindow.cpp [hunk 10]
```diff
@@ -232,25 +237,55 @@ void MainWindow::resizeEvent(QResizeEvent* event) {
  * @param newPos current position of the slider
```
Comment (parso619) [inline ViAn/GUI/mainwindow.cpp]: Maybe update this comment so that it describes more about what is dane in the function.

File: ViAn/GUI/mainwindow.cpp (hunk 8 above)
Comment (henpe071) [inline ViAn/GUI/mainwindow.cpp]: Remove print

File: ViAn/GUI/mainwindow.cpp [hunk 11]
```diff
@@ -232,25 +237,55 @@ void MainWindow::resizeEvent(QResizeEvent* event) {
...
+            emit set_pause_video();
+            was_paused = true;
+        }*/
+
+        std::chrono::milliseconds current_time = std::chrono::duration_cast<
+                std::chrono::milliseconds >(
+                    std::chrono::system_clock::now().time_since_epoch()
+        );
+        std::chrono::milliseconds time_since_last_slider_frame_update = current_time-slider_timer;
+        if (time_since_last_slider_frame_update.count() >= 200) {
```
Comment (parso619) [inline ViAn/GUI/mainwindow.cpp]: If 200 is a constant value, maybe it's better to add it as a const variable instead.

Comment (parso619) [inline ViAn/GUI/mainwindow.cpp]: If 200 is a constant value, maybe it's better to add it as a const variable instead.

File: ViAn/GUI/mainwindow.cpp (hunk 9 above)
Comment (henpe071) [inline ViAn/GUI/mainwindow.cpp]: Remove print

File: ViAn/GUI/mainwindow.h [hunk 12]
```diff
@@ -132,9 +137,20 @@ private slots:
...
     IconOnButtonHandler *iconOnButtonHandler;
-
     QSlider *video_slider;
 
+    bool slider_blocked = false;
+    bool slider_moving = false;
+    bool slider_paused_video = false;
+    int prev_slider_pos = 0;
+
+    int posCounter = 0;
```
Comment (henpe071) [inline ViAn/GUI/mainwindow.h]: case

File: ViAn/Test/test_video_player.cpp [hunk 13]
```diff
@@ -246,3 +226,17 @@ void test_video_player::test_set_stop_video() {
     QVERIFY(!mvideo->video_paused);
     QVERIFY(mvideo->video_stopped);
 }
+
+/**
+ * @brief test_video_player::test_on_set_playback_fram
+ */
+void test_video_player::test_on_set_playback_fram() {
```
Comment (parso619) [inline ViAn/Test/test_video_player.cpp]: Misspelled: `test_on_set_playback_frame`

File: ViAn/GUI/mainwindow.cpp [hunk 14]
```diff
@@ -232,25 +237,55 @@ void MainWindow::resizeEvent(QResizeEvent* event) {
  * @param newPos current position of the slider
  */
 void MainWindow::on_videoSlider_valueChanged(int newPos){
```
Comment (GeorgeYildiz) [inline ViAn/GUI/mainwindow.cpp]: Refactor this to make it easier to read. If the method does more than one thing, separate based on those.

**Comment by parso619** [comment]
This code does not seem to be up to date with master. Several of the functions that are removed have already been removed in previous pull requests. For example `test_stop_video` and `stop_video`.

File: ViAn/Test/test_video_player.cpp (hunk 13 above)
Comment (GeorgeYildiz) [inline ViAn/Test/test_video_player.cpp]: frame is misspelled in the method name "fram". Also it could be good to test a frame number that is negative or larger than the total amounts of frames in the video.

**Comment by fredrikiselius** [comment]
Still contains bugs. Dont merge
